### PR TITLE
fix(major): [sc-10387] Try to reconnect to service while it is in Consul.

### PR DESCRIPTION
**Pull request by ser-0xff**
## Description

Try to reconnect to service while it is in Consul.
API has not been changed, but changes are quite sensitive so let's make it as a major to be sure dependent projects will not be automatically switched to the new version.

## How Has This Been Tested?

Unit tests. 

## Minimal checklist:

- [X] I have performed a self-review of my own code 
- [ ] I have added `DocC` code-level documentation for any public interfaces exported by the package
- [ ] I have added unit and/or integration tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by dukhnyak — APPROVED** [review]

**Review by dimlio — APPROVED** [review]

File: Sources/DistributedSystem/DiscoveryManager.swift [hunk 1]
```diff
@@ -16,7 +16,24 @@ import PackageConcurrencyHelpers
 final class DiscoveryManager {
     private final class ProcessInfo {
         var channel: (UInt32, Channel)?
-        var pendingServices = [(NodeService, DistributedSystem.ConnectionHandler)]()
+        var pendingHandlers = [String: [DistributedSystem.ConnectionHandler]]()
+        var connectedServices = Set<String>()
+
+        func addPendingHandlers(_ serviceName: String, _ handlers: [DistributedSystem.ConnectionHandler]) {
+            var pendingHandlers = self.pendingHandlers[serviceName, default: []]
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DiscoveryManager.swift]: Fixed.

File: Sources/DistributedSystem/DiscoveryManager.swift [hunk 2]
```diff
@@ -16,7 +16,24 @@ import PackageConcurrencyHelpers
...
+            pendingHandlers.append(contentsOf: handlers)
+            self.pendingHandlers[serviceName] = pendingHandlers
+        }
+
+        func setChannel(_ channelID: UInt32, _ channel: Channel) -> [String: [DistributedSystem.ConnectionHandler]] {
+            self.channel = (channelID, channel)
+            for serviceName in pendingHandlers.keys {
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DiscoveryManager.swift]: Fixed.

File: Tests/DistributedSystemTests/DistributedSystemTests.swift [hunk 3]
```diff
@@ -294,6 +294,67 @@ final class DistributedSystemTests: XCTestCase {
     }
     */
 
+    func testReconnect() async throws {
+        // Checking the distributed actors do not leak,
+        // use a closure here to be sure at the check point all references will be released
+        let flags = Flags()
+        _ = try await {
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Tests/DistributedSystemTests/DistributedSystemTests.swift]: Fixed.

File: Sources/DistributedSystem/DiscoveryManager.swift [hunk 4]
```diff
@@ -314,26 +359,70 @@ final class DiscoveryManager {
...
+
+            for serviceName in processInfo.connectedServices {
+                guard let discoveryInfo = self.discoveries[serviceName] else {
+                    logger.error("internal error: service \(serviceName) not found")
+                    continue
+                }
+
+                var servicePendingHandlers = [DistributedSystem.ConnectionHandler]()
+
+                for serviceInfo in discoveryInfo.services.values {
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DiscoveryManager.swift:399]: That fix change nothing.

File: Sources/DistributedSystem/DiscoveryManager.swift [hunk 5]
```diff
@@ -314,26 +359,70 @@ final class DiscoveryManager {
                     }
                 }
             }
+
+            if stop {
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DiscoveryManager.swift]: Well, we can...

**Comment by dukhnyak** [comment]
I am happy to give a try to it locally on my computer once PR is ready

**Pull request merged**